Content.Client/Stack/StackSystem.cs

```diff
@@ -56,20 +56,25 @@ private void OnAppearanceChange(EntityUid uid, StackComponent comp, ref Appearan
             if (args.Sprite == null || comp.LayerStates.Count < 1)
                 return;
 
+            StackLayerData data = new StackLayerData(); // Frontier: use structure to store StackLayerData
+
             // Skip processing if no actual
-            if (!_appearanceSystem.TryGetData<int>(uid, StackVisuals.Actual, out var actual, args.Component))
+            if (!_appearanceSystem.TryGetData<int>(uid, StackVisuals.Actual, out data.Actual, args.Component))
                 return;
 
-            if (!_appearanceSystem.TryGetData<int>(uid, StackVisuals.MaxCount, out var maxCount, args.Component))
-                maxCount = comp.LayerStates.Count;
+            if (!_appearanceSystem.TryGetData<int>(uid, StackVisuals.MaxCount, out data.MaxCount, args.Component))
+                data.MaxCount = comp.LayerStates.Count;
+
+            if (!_appearanceSystem.TryGetData<bool>(uid, StackVisuals.Hide, out data.Hidden, args.Component))
+                data.Hidden = false;
 
-            if (!_appearanceSystem.TryGetData<bool>(uid, StackVisuals.Hide, out var hidden, args.Component))
-                hidden = false;
+            if (comp.LayerFunction is not null) // Frontier: use stack layer function to modify appearance if provided.
+                comp.LayerFunction.Apply(ref data); // Frontier
```

Comment:
instead, you should just add a bool flag to the stack component to check for that instead, and then call the shared function here after bringing it in as a dependency.